## Assignment 4

In [2]:
import pandas as pd
import numpy as np

#vix
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import OneHotEncoder

#load StatsModel API
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [2]:
conda install plotly

Retrieving notices: ...working... DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:https://repo.anaconda.com:443 "GET /pkgs/main/notices.json HTTP/1.1" 404 None
DEBUG:urllib3.connectionpool:https://repo.anaconda.com:443 "GET /pkgs/r/notices.json HTTP/1.1" 404 None
done
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): repo.anaconda.com:443
- DEBUG:urllib3.connectionpool:https://repo.anaconda.com:443 "GET /pkgs/main/noarch/current_repodata.json HTTP/1.1" 200 None
DEBUG:urllib3.connectionpool:https://repo.anaconda.com:443 "GET /pkgs/r/noarch/current_repodata.json HTTP/1.1" 304 0
DEBUG:urllib3.connectionpool:https://repo.anaconda.com:443 "GET /pkgs/main/osx

In [3]:
from sklearn import datasets

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import KFold, cross_val_score

### 3. We now review k-fold cross-validation.

#### (a) Explain how k-fold cross-validation is implemented.

* K-fold cross-validation is a method for assessing the performance of a machine learning model and mitigating overfitting by iteratively splitting the dataset into k equally sized subsets.

#### (b) What are the advantages and disadvantages of k-fold cross-validation relative to:
####  i. The validation set approach?
* Advantages of k-fold cross-validation:
Better utilization of data: K-fold cross-validation utilizes the entire dataset for both training and validation, providing a more comprehensive assessment of the model's performance.
Robust performance estimate: By averaging performance metrics across multiple folds, k-fold cross-validation produces a more reliable estimate of the model's generalization ability.
Reduced variance: The variance of the performance estimate is lower as it's averaged over multiple validation sets.

* Disadvantages of k-fold cross-validation relative to the validation set approach:
Computational overhead: K-fold cross-validation requires training and evaluating the model k times, which can be computationally expensive, especially for large datasets or complex models.
Increased training time: Splitting the dataset into multiple folds and training the model repeatedly can prolong the training process.
####  ii. LOOCV?
* Advantages of k-fold cross-validation:
Reduced bias: K-fold cross-validation typically has lower bias compared to LOOCV because it averages over multiple validation sets, reducing the impact of a single data point on the model's evaluation.
Less sensitive to outliers: K-fold cross-validation is less affected by outliers compared to LOOCV, as it aggregates results over multiple folds.

* Disadvantages of k-fold cross-validation relative to LOOCV:
Higher variance: K-fold cross-validation may have higher variance compared to LOOCV, especially when the dataset is small, as it involves random partitioning of data into folds.
Computational complexity: Although LOOCV involves training the model only once per data point, k-fold cross-validation requires training the model multiple times, leading to increased computational complexity, especially for large datasets.

### 5. In Chapter 4, we used logistic regression to predict the probability of default using income and balance on the Default data set. We will now estimate the test error of this logistic regression model using the validation set approach. Do not forget to set a random seed before beginning your analysis.
#### (a) Fit a logistic regression model that uses income and balance to predict default.

In [4]:
default= pd.read_csv('Default.csv', na_values='?').dropna()

# Display the first few rows
default.head() 

default student      balance        income
0      No      No   729.526495  44361.625074
1      No     Yes   817.180407  12106.134700
2      No      No  1073.549164  31767.138947
3      No      No   529.250605  35704.493935
4      No      No   785.655883  38463.495879

In [5]:
default.describe

<bound method NDFrame.describe of      default student      balance        income
0         No      No   729.526495  44361.625074
1         No     Yes   817.180407  12106.134700
2         No      No  1073.549164  31767.138947
3         No      No   529.250605  35704.493935
4         No      No   785.655883  38463.495879
...      ...     ...          ...           ...
9995      No      No   711.555020  52992.378914
9996      No      No   757.962918  19660.721768
9997      No      No   845.411989  58636.156984
9998      No      No  1569.009053  36669.112365
9999      No     Yes   200.922183  16862.952321

[10000 rows x 4 columns]>

In [29]:
# Assuming 'default' and 'student' columns contain categorical values ('yes' or 'no')

# Mapping 'yes' to 1 and 'no' to 0
default['default'] = default['default'].map({'Yes': 1, 'No': 0})
default['student'] = default['student'].map({'Yes': 1, 'No': 0})

# Or using replace method
default['default'] = default['default'].replace({'Yes': 1, 'No': 0})
default['student'] = default['student'].replace({'Yes': 1, 'No': 0})

default['default'] = default['default'].astype('int64')
default['student'] = default['student'].astype('int64')
default['income'] = default['income'].astype('int64')
default['income'] = default['income'].astype('int64')
default.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   default  10000 non-null  int64
 1   student  10000 non-null  int64
 2   balance  10000 non-null  int64
 3   income   10000 non-null  int64
dtypes: int64(4)
memory usage: 312.6 KB


In [31]:
from sklearn.model_selection import train_test_split

# Convert the 'default' column to int64
endog = Default.astype("int64")

endog

0       0
1       0
2       0
3       0
4       0
       ..
9995    0
9996    0
9997    0
9998    0
9999    0
Name: default, Length: 10000, dtype: int64

In [32]:
exog = sm.add_constant(default.drop(columns=["student"]))
exog

const  default  balance  income
0       1.0        0      729   44361
1       1.0        0      817   12106
2       1.0        0     1073   31767
3       1.0        0      529   35704
4       1.0        0      785   38463
...     ...      ...      ...     ...
9995    1.0        0      711   52992
9996    1.0        0      757   19660
9997    1.0        0      845   58636
9998    1.0        0     1569   36669
9999    1.0        0      200   16862

[10000 rows x 4 columns]

In [33]:
# Now fit the logistic regression model
logit_mod = sm.Logit(endog, exog)
logit_result = logit_mod.fit()
logit_result.summary()

         Current function value: 0.000000
         Iterations: 35


/Users/nupoor/anaconda3/lib/python3.11/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
/Users/nupoor/anaconda3/lib/python3.11/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
/Users/nupoor/anaconda3/lib/python3.11/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSeparationWarning)
/Users/nupoor/anaconda3/lib/python3.11/site-packages/statsmodels/discrete/discrete_model.py:227: PerfectSeparationWarning: Perfect separation or prediction detected, parameter may not be identified
  warnings.warn(msg, category=PerfectSep

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                default   No. Observations:                10000
Model:                          Logit   Df Residuals:                     9996
Method:                           MLE   Df Model:                            3
Date:                Thu, 07 Mar 2024   Pseudo R-squ.:                   1.000
Time:                        19:34:11   Log-Likelihood:            -1.9824e-06
converged:                      False   LL-Null:                       -1460.3
Covariance Type:            nonrobust   LLR p-value:                     0.000
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -19.4157   2236.895     -0.009      0.993   -4403.648    4364.817
default       45.1131   3337.179      0.014      0.989   -6495.638    6585.865
balance       -0.0008      1.736     -0.000      1.000      -3.404       3.402
income        -0.0001      0.073     -0.001      0.999      -0.143       0.143
==============================================================================

Complete Separation: The results show that there iscomplete separation or perfect prediction.
In this case the Maximum Likelihood Estimator does not exist and the parameters
are not identified.
"""

#### (b) Using the validation set approach, estimate the test error of this model. In order to do this, you must perform the following steps:
#### i. Split the sample set into a training set and a validation set.
#### ii. Fit a multiple logistic regression model using only the training observations.
#### iii. Obtain a prediction of default status for each individual in the validation set by computing the posterior probability of default for that individual, and classifying the individual to the default category if the posterior probability is greater than 0.5.
#### iv. Compute the validation set error, which is the fraction of the observations in the validation set that are misclassified.

In [41]:
from ISLP.models import (ModelSpec as MS,
                         summarize ,
                         poly)
from functools import partial
from sklearn.model_selection import \
(cross_validate , KFold , ShuffleSplit)
from sklearn.base import clone
from ISLP.models import sklearn_sm


random_seed=1
default_train, default_valid = train_test_split(default,
                                             test_size=5000, random_state=random_seed)

dd = MS(['balance', 'income'])
X_train = dd.fit_transform(default_train) 
y_train = default_train['default']
model = sm.OLS(y_train, X_train)
results = model.fit()

X_valid = dd.transform(default_valid) 
y_valid = default_valid['default'] 
valid_pred = results.predict(X_valid)
np.mean((y_valid - valid_pred)**2)

0.027289082195555328

#### (c) Repeat the process in (b) three times, using three different splits of the observations into a training set and a validation set. Comment on the results obtained.

In [47]:
default_train, default_valid = train_test_split(default,
                                             test_size=2000, random_state=random_seed)

dd = MS(['balance', 'income'])
X_train = dd.fit_transform(default_train) 
y_train = default_train['default']
model = sm.OLS(y_train, X_train)
results = model.fit()

X_valid = dd.transform(default_valid) 
y_valid = default_valid['default'] 
valid_pred = results.predict(X_valid)
np.mean((y_valid - valid_pred)**2)

0.025981211117436117

In [48]:
default_train, default_valid = train_test_split(default,
                                             test_size=3500, random_state=random_seed)

dd = MS(['balance', 'income'])
X_train = dd.fit_transform(default_train) 
y_train = default_train['default']
model = sm.OLS(y_train, X_train)
results = model.fit()

X_valid = dd.transform(default_valid) 
y_valid = default_valid['default'] 
valid_pred = results.predict(X_valid)
np.mean((y_valid - valid_pred)**2)

0.026630463503038455

In [50]:
default_train, default_valid = train_test_split(default,
                                             test_size=4000, random_state=random_seed)

dd = MS(['balance', 'income'])
X_train = dd.fit_transform(default_train) 
y_train = default_train['default']
model = sm.OLS(y_train, X_train)
results = model.fit()

X_valid = dd.transform(default_valid) 
y_valid = default_valid['default'] 
valid_pred = results.predict(X_valid)
np.mean((y_valid - valid_pred)**2)

0.026646818484376566

* MSE is least in 2000(0.02598) as comapred to other tested sizes such as 5000, 3500, and 4000, underscores the importance of considering the trade-offs between model performance and computational efficiency. It also suggests that the model's predictive performance was optimal or most accurate when evaluated on datasets of this size. While larger test data sizes might theoretically provide a more robust evaluation of the model's generalization ability, the decrease in MSE with a test size of 2000 suggests that there might be a point of diminishing returns, where additional data beyond a certain threshold does not significantly improve predictive performance.

#### (d) Now consider a logistic regression model that predicts the prob- ability of default using income, balance, and a dummy variable for student. Estimate the test error for this model using the val- idation set approach. Comment on whether or not including a dummy variable for student leads to a reduction in the test error rate.

In [56]:
# Define endog and exog variables
y = default['default']
X = default[['balance', 'income', 'student']]

# Add a constant term to the independent variables
X = sm.add_constant(X)

# Convert the 'student' column into a dummy variable
dummy_student = pd.get_dummies(default['student'], drop_first=True)

# Concatenate the dummy variable with the independent variables
X = pd.concat([X, dummy_student], axis=1)

default

default  student  balance  income
0           0        0      729   44361
1           0        1      817   12106
2           0        0     1073   31767
3           0        0      529   35704
4           0        0      785   38463
...       ...      ...      ...     ...
9995        0        0      711   52992
9996        0        0      757   19660
9997        0        0      845   58636
9998        0        0     1569   36669
9999        0        1      200   16862

[10000 rows x 4 columns]

In [57]:
# Fit the logistic regression model
logit_mod = sm.Logit(y, X)
logit_result = logit_mod.fit()

# Display the summary of the model
print(logit_result.summary())


Optimization terminated successfully.
         Current function value: 0.078581
         Iterations 10
                           Logit Regression Results                           
Dep. Variable:                default   No. Observations:                10000
Model:                          Logit   Df Residuals:                     9996
Method:                           MLE   Df Model:                            3
Date:                Thu, 07 Mar 2024   Pseudo R-squ.:                  0.4619
Time:                        20:15:10   Log-Likelihood:                -785.81
converged:                       True   LL-Null:                       -1460.3
Covariance Type:            nonrobust   LLR p-value:                3.389e-292
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -10.8658      0.492    -22.078      0.000     -11.830      -9.901
balance        0.0057      0

In [63]:
default_train, default_valid = train_test_split(default,
                                             test_size=2000, random_state=random_seed)

dd = MS(['balance', 'income', 'student'])
X_train = dd.fit_transform(default_train) 
y_train = default_train['default']
model = sm.OLS(y_train, X_train)
results = model.fit()

X_valid = dd.transform(default_valid) 
y_valid = default_valid['default'] 
valid_pred = results.predict(X_valid)
np.mean((y_valid - valid_pred)**2)

0.025985810872869025

### - Creating a dummy variable does not reduce the test error.

### 6. We continue to consider the use of a logistic regression model to predict the probability of default using income and balance on the Default data set. In particular, we will now compute estimates for the standard errors of the income and balance logistic regression coeffi- cients in two different ways: (1) using the bootstrap, and (2) using the standard formula for computing the standard errors in the sm.GLM() function. Do not forget to set a random seed before beginning your analysis.
#### (a) Using the summarize() and sm.GLM() functions, determine the estimated standard errors for the coefficients associated with income and balance in a multiple logistic regression model that uses both predictors.

In [64]:
# Set a random seed
random_seed = 42

# Define endog and exog variables
endog = default['default']
exog = default[['income', 'balance']]

# Add a constant term to the independent variables
exog = sm.add_constant(exog)

# Fit the logistic regression model
logit_model = sm.GLM(endog, exog, family=sm.families.Binomial())
logit_result = logit_model.fit()

# Display the summary of the model
print(logit_result.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:                default   No. Observations:                10000
Model:                            GLM   Df Residuals:                     9997
Model Family:                Binomial   Df Model:                            2
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -789.52
Date:                Thu, 07 Mar 2024   Deviance:                       1579.0
Time:                        20:21:48   Pearson chi2:                 6.95e+03
No. Iterations:                     9   Pseudo R-squ. (CS):             0.1256
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -11.5370      0.435    -26.544      0.0

#### (b) Write a function, boot_fn(), that takes as input the Default data set as well as an index of the observations, and that outputs the coefficient estimates for income and balance in the multiple logistic regression model.

In [75]:
def boot_fn(data, index, formula=None):
    if formula:
        model = sm.GLM.from_formula(formula, family=sm.families.Binomial(), data=data.iloc[index])
    else:
        endog = data['default'].iloc[index]
        exog = data[['income', 'balance']].iloc[index]
        exog = sm.add_constant(exog)
        model = sm.GLM(endog, exog, family=sm.families.Binomial())
    result = model.fit()
    return result.params

def bootstrap(data, n_iterations=1000, formula=None):
    bootstrap_coefs = []
    for _ in range(n_iterations):
        index = np.random.choice(data.index, size=len(data), replace=True)
        bootstrap_coefs.append(boot_fn(data, index, formula=formula))

    bootstrap_coefs = pd.DataFrame(bootstrap_coefs, columns=['Intercept', 'income', 'balance'])
    bootstrap_se = bootstrap_coefs.std()
    return bootstrap_se  # Add return statement here

# Example usage:
# Assuming 'default' is your DataFrame
bootstrap_se = bootstrap(default)
print(bootstrap_se)


Intercept         NaN
income       0.000005
balance      0.000227
dtype: float64


#### as this does not work well. Trying the below one

In [79]:
def boot_fn(default, index):
    model = sm.GLM.from_formula('default ~ income + balance', family=sm.families.Binomial(), data=default.iloc[index])
    result = model.fit()
    return result.params

n_iterations = 1000
bootstrap_coefs = []
for _ in range(n_iterations):
    index = np.random.choice(default.index, size=len(default), replace=True)
    bootstrap_coefs.append(boot_fn(default, index))

bootstrap_coefs = pd.DataFrame(bootstrap_coefs, columns=['Intercept', 'income', 'balance'])
# Calculate standard errors
bootstrap_se = bootstrap_coefs.std()
bootstrap_se

Intercept    0.429394
income       0.000005
balance      0.000229
dtype: float64

In [140]:
def boot_fn(default, index):
    # Select subset of data based on the given index
    subset = default.iloc[index]
    
    # Extract features and target variable
    X_boot = subset[['income', 'balance']]
    y_boot = subset['default']
    
    # Fit logistic regression model
    model_boot = sm.Logit(y_boot, X_boot)
    result_boot = model_boot.fit(disp=0)
    
    # Return coefficient estimates for 'income' and 'balance'
    return result_boot.params

# Example usage:
# Assuming 'default' is your dataset and 'index' is a list of indices for the subset of data
coeff = boot_fn(default, index)
coeff

income    -0.000129
balance    0.000477
dtype: float64

#### (c) Following the bootstrap example in the lab, use your boot_fn() function to estimate the standard errors of the logistic regression coefficients for income and balance.

In [138]:
def alpha_func(D, idx):
    cov_ = np.cov(D[['income', 'balance']].iloc[idx], rowvar=False)
    return ((cov_[1,1] - cov_[0,1]) / (cov_[0,0] + cov_[1,1] - 2*cov_[0,1]))

def boot_SE(func, D, n=None, B=1000, seed=0):
    rng = np.random.default_rng(seed)
    first_, second_ = 0, 0
    n = n or D.shape[0]
    for _ in range(B):
        idx = rng.choice(D.index, n, replace=True)
        value = func(D, idx)
        first_ += value
        second_ += value**2
    return np.sqrt(second_ / B - (first_ / B)**2)

# Assuming 'default' is your dataset with predictor variables 'income', 'balance', 'student', and target variable 'default'
alpha_SE = boot_SE(alpha_func, default, B=1000, seed=0)
print("Standard Error (alpha):", alpha_SE)

Standard Error (alpha): 0.00033958238300135443


#### (d) Comment on the estimated standard errors obtained using the sm.GLM() function and using the bootstrap.

* The standard errors obtained viz; income(-0.000129), balance(0.000477) and overall standard error(0.00033958238300135443) from the bootstrap tend to provide more robust estimates of the variability of the coefficient estimates compared to those obtained from the sm.GLM() function. Bootstrap resampling captures the variability in the data more accurately, making it a valuable tool for assessing the uncertainty of parameter estimates when analytical methods might not fully account for all sources of variability.

* Tried both sm.glm and logit functions.

### 9. We will now consider the Boston housing data set, from the ISLP library.
#### (a) Based on this data set, provide an estimate for the population mean of medv. Call this estimate μˆ.

In [112]:
# Load the Boston dataset
boston_file = "https://raw.githubusercontent.com/selva86/datasets/master/BostonHousing.csv"
boston = pd.read_csv(boston_file)


mu_hat = boston['medv'].mean()

print("Estimated mean of medv (μˆ):", mu_hat)

Estimated mean of medv (μˆ): 22.532806324110677


In [114]:
boston.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        b  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

#### (b) Provide an estimate of the standard error of μˆ. Interpret this result.
#### Hint: We can compute the standard error of the sample mean by dividing the sample standard deviation by the square root of the number of observations.

In [113]:
sample_std = np.std(boston['medv'], ddof=1)

# Calculate the number of observations
n = len(boston['medv'])

# Calculate the standard error of the sample mean
standard_error_mu_hat = sample_std / np.sqrt(n)

print("Standard error of μˆ:", standard_error_mu_hat)

Standard error of μˆ: 0.4088611474975351


#### (c) Now estimate the standard error of μˆ using the bootstrap. How does this compare to your answer from (b)?

In [118]:
# Define function to calculate sample mean
def sample_mean(boston):
    return np.mean(boston)

# Define function for bootstrapping
def bootstrap(boston, func, n_iterations=1000):
    sample_means = []
    for _ in range(n_iterations):
        # Resample data with replacement
        resampled_data = np.random.choice(boston, size=len(boston), replace=True)
        # Calculate sample mean
        sample_means.append(func(resampled_data))
    return np.std(sample_means)

# Calculate standard error of μˆ using bootstrap
bootstrap_se = bootstrap(boston['medv'], sample_mean)

print("Standard error of μˆ using bootstrap:", bootstrap_se)

Standard error of μˆ using bootstrap: 0.4023244298640598


- The standard error using bootstrap is slightly lower than without bootstrap method. The slightly lower standard error obtained with the bootstrap method suggests that resampling from the dataset helps to capture the variability more accurately, resulting in a more precise estimate of the standard error. This improvement in precision can be attributed to the bootstrap's ability to account for the inherent randomness in the dataset and provide a robust estimate of the standard error without relying on assumptions about the underlying distribution. It is more reliable and less susceptible to sampling variability. 

#### (d) Based on your bootstrap estimate from (c), provide a 95 % confidence interval for the mean of medv. Compare it to the results obtained by using Boston['medv'].std() and the two standard error rule (3.9).
#### Hint: You can approximate a 95 % confidence interval using the formula [μˆ − 2SE(μˆ), μˆ + 2SE(μˆ)].

In [120]:
bootstrap_se = boot_SE(lambda D, idx: np.mean(D.loc[idx, 'medv']), boston, B=1000, seed=0)

# Compute the standard error of the mean (SEM)
SEM = bootstrap_se / np.sqrt(len(boston))

# Calculate the confidence interval
confidence_interval = (np.mean(boston['medv']) - 2 * SEM, np.mean(boston['medv']) + 2 * SEM)

confidence_interval

(22.49612750938295, 22.569485138838402)

#### (e) Based on this data set, provide an estimate, μˆmed, for the median value of medv in the population.

In [121]:
median_medv = np.median(boston['medv'])

median_medv

21.2

#### (f) We now would like to estimate the standard error of μˆmed. Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings.

In [123]:
def bootstrap_median(boston, B=1000, seed=0):
    rng = np.random.default_rng(seed)
    medians = []
    n = len(boston)
    
    for _ in range(B):
        # Generate bootstrap sample by resampling with replacement
        bootstrap_sample = rng.choice(boston, size=n, replace=True)
        # Calculate median of bootstrap sample
        median = np.median(bootstrap_sample)
        medians.append(median)
    
    # Compute standard error of the median
    std_error = np.std(medians, ddof=1)
    return std_error

# Assuming 'boston' is your DataFrame and 'CMEDV' is the column containing median values
std_error_median = bootstrap_median(boston['medv'])

std_error_median

0.36963108248243537

- The standard error of the median, estimated using the bootstrap method, is approximately 0.3696. This value represents the variability or uncertainty associated with our estimate of the median value of medv in the population. A lower standard error indicates higher precision in our estimate, meaning that our median estimate is likely to be closer to the true population median.

#### (g) Based on this data set, provide an estimate for the tenth per- centile of medv in Boston census tracts. Call this quantity μˆ0.1. (You can use the np.percentile() function.) np.

In [126]:
tenth_percentile = np.percentile(boston['medv'], 10)
tenth_percentile

12.75

#### (h) Use the bootstrap to estimate the standard error of μˆ0.1. Comment on your findings.

In [128]:
def tenth_percentile_func(D, idx):
    return np.percentile(D.loc[idx, 'medv'], 10)

# Define a function to calculate the standard error of the tenth percentile using the bootstrap method
def boot_SE_tenth_percentile(func, D, n=None, B=1000, seed=0):
    rng = np.random.default_rng(seed)
    percentile_values = []
    n = n or D.shape[0]
    for _ in range(B):
        idx = rng.choice(D.index, n, replace=True)
        percentile_value = func(D, idx)
        percentile_values.append(percentile_value)
    return np.std(percentile_values, ddof=1)

# Estimate the standard error of μˆ0.1 using the bootstrap method
bootstrap_se_tenth_percentile = boot_SE_tenth_percentile(tenth_percentile_func, boston, B=1000, seed=0)
bootstrap_se_tenth_percentile

0.5037060251374404

- The error is highest that is 0.5037 as compared to Standard error of μˆ: 0.4088, Standard error of μˆ using bootstrap: 0.4023 and std_error_median 0.36963108248243537 indicating more variability in the estimate of the tenth percentile compared to the mean. It may not necessarily outperform direct calculations using percentile functions in terms of precision. However, it still offers a robust and flexible approach, especially when exact formulas for standard errors are not readily available.